### ISSUE TITLE
Pluggable Qtile

### ISSUE THREAD

**Issue by cblegare** [qtile/qtile#1073]
Hi!

**tl;dr** any thoughts about a plugin API for Qtile?

---

I started using Qtile earlier this year and I love it, thanks :).  I did [some work][1] for wrapping dmenu with python, then thought about ways of integrating it into qtile. I stumbled on the new [extension API][2] and found nothing public (i.e. not starting with an `_`) I could use for this stuff.  

Also, I felt that all those extensions could grow very large: integration of third party softwares from within the Qtile source base.  I remember reading somewhere a discussion you guys had about having the core libqtile somewhere and "clients" (widgets, layouts, etc) elsewhere (maybe completely separate projects?).

I have some experience with the [entrypoint API][3] provided by `setuptools`.  I think that [`pkg_resources` entry point API][4] could be very interesting in this case.  For instance

* `qtile.widget`
* `qtile.layout`
* `qtile.command`

could all be entry point group names that qtile defines.

I wonder if this issue was ever discussed.  I prefer asking for insights before diving into code+pull requests, because this kind of API involves a deep understanding of the lifecycle of the project, which clearly I don't have.

Any thoughts?

[1]: http://xdmenu.readthedocs.io/en/latest/index.html
[2]: https://github.com/qtile/qtile/tree/develop/libqtile/extension
[3]: http://setuptools.readthedocs.io/en/latest/setuptools.html#dynamic-discovery-of-services-and-plugins
[4]: http://setuptools.readthedocs.io/en/latest/pkg_resources.html#entry-points

**Comment by zordsdavini** [comment]
I love Qtile integration with 3rd party software. I did some tricks on dmenu, moc. There are a lot of commands starting with `cmd_*` in https://github.com/qtile/qtile/blob/develop/libqtile/manager.py . These can be reached in your outside script. Check https://github.com/qtile/qtile-examples/blob/master/zordsdavini/bin/dmenu-qtile-windowlist.py . Now it is integrated with dmenu extenssions. But  _entrypoint API_ looks very interesting.

**Comment by cblegare** [comment]
Thank you for the links!

The [Command API][1] lets me control Qtile from keys and external scripts; python scripts using the Client singleton provider or shell script using Qsh.

With [Keys][2], I can also use Qtile to control other piece of software in subprocesses. 

I though of using Qtile to hold my python module in it's process (i.e. import from the config.py then bind a function call to a Key),  but I understand that this may not be a good idea: I should let the Qtile process do it's windowing stuff and keep my own processes somewhere else.  If my python code takes too much time to start (because it reads many files on initialization) I should consider caching or daemons.

---

I wish that `cmd_*` stuff was documented.  I may try and fine some time to request a pull on this matter!

Still, entrypoints could be of some use in case built-ins were to be moved outside of libqtile (which looks like a good idea for built-in Widgets, Layouts, etc, but this is only my opinion).

So this is a [`docs`](https://github.com/qtile/qtile/labels/docs) issue.

[1]: http://docs.qtile.org/en/latest/manual/commands/index.html
[2]: http://docs.qtile.org/en/latest/manual/config/keys.html

**Comment by kynikos** [comment]
Hi, it's going to be great if you find a way to document the `cmd_*` calls! The discussion about splitting the bar and the other extensions to different processes is #622, and that would surely depend on a more thorough qtile API documentation.

**Comment by m-col** [comment]
Looks like this is ultimately a duplicate of https://github.com/qtile/qtile/issues/1360, so closing this. It would indeed be great to have those commands all documented!
